File: frontend/demo/component/timepicker/time-picker-custom-validation.ts
import 'Frontend/demo/init'; // hidden-source-line

import { html, LitElement } from 'lit';
import { customElement } from 'lit/decorators.js';
import '@vaadin/time-picker';
import { Binder, field } from '@hilla/form';
import AppointmentModel from 'Frontend/generated/com/vaadin/demo/domain/AppointmentModel';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('time-picker-custom-validation')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  // tag::snippet[]
  private binder = new Binder(this, AppointmentModel);

  protected override firstUpdated() {
    this.binder.for(this.binder.model.startTime).addValidator({
      message: 'The selected time is not available',
      validate: (startTime: string) =>
        (startTime >= '08:00' && startTime <= '12:00') ||
        (startTime >= '13:00' && startTime <= '16:00'),
    });
  }

  protected override render() {
    return html`
      <vaadin-time-picker
        label="Appointment time"
        helper-text="Open 8:00-12:00, 13:00-16:00"
        min="08:00"
        max="16:00"
        .step="${60 * 30}"
        ${field(this.binder.model.startTime)}
      ></vaadin-time-picker>
    `;
  }
  // end::snippet[]
}

File: src/main/java/com/vaadin/demo/component/timepicker/TimePickerCustomValidation.java
package com.vaadin.demo.component.timepicker;

import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.timepicker.TimePicker;
import com.vaadin.flow.data.binder.Binder;
import com.vaadin.flow.router.Route;

import java.time.Duration;
import java.time.LocalTime;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.domain.Appointment;

@Route("time-picker-custom-validation")
public class TimePickerCustomValidation extends Div {

    public TimePickerCustomValidation() {
        // tag::snippet[]
        TimePicker timePicker = new TimePicker();
        timePicker.setLabel("Appointment time");
        timePicker.setHelperText("Open 8:00-12:00, 13:00-16:00");
        timePicker.setStep(Duration.ofMinutes(30));
        timePicker.setMinTime(LocalTime.of(8, 0));
        timePicker.setMaxTime(LocalTime.of(16, 0));
        add(timePicker);

        Binder<Appointment> binder = new Binder<>(Appointment.class);
        binder.forField(timePicker).withValidator(startTime -> {
            return !(LocalTime.of(8, 0).isAfter(startTime)
                    || (LocalTime.of(12, 0).isBefore(startTime) && LocalTime.of(13, 0).isAfter(startTime))
                    || LocalTime.of(16, 0).isBefore(startTime));
        }, "The selected time is not available").bind(Appointment::getStartTime, Appointment::setStartTime);
        // end::snippet[]
    }

    public static class Exporter extends DemoExporter<TimePickerCustomValidation> { // hidden-source-line
    } // hidden-source-line
}
